File: go/mechanisms/evm/verify_erc6492.go
package evm

import (
	"context"

	"github.com/ethereum/go-ethereum/common"
)

// VerifyERC6492Signature verifies an ERC-6492 counterfactual signature by calling the
// ERC-6492 UniversalSigValidator contract via eth_call (no state changes committed).
// The validator atomically simulates the factory deployment then verifies the inner
// signature using EIP-1271 isValidSignature on the resulting contract.
//
// Returns false (not an error) if the validator returns false.
// Returns false + error if the validator contract is unavailable or the call fails.
func VerifyERC6492Signature(
	ctx context.Context,
	facilitatorSigner FacilitatorEvmSigner,
	signerAddress string,
	hash [32]byte,
	signature []byte,
) (bool, error) {
	signerAddr := common.HexToAddress(signerAddress)
	result, err := facilitatorSigner.ReadContract(
		ctx,
		UniversalSigValidatorAddress,
		UniversalSigValidatorABI,
		"isValidSig",
		signerAddr,
		hash,
		signature,
	)
	if err != nil {
		return false, err
	}
	valid, ok := result.(bool)
	if !ok {
		return false, nil
	}
	return valid, nil
}

File: typescript/packages/mechanisms/evm/src/verify_erc6492.ts
import { Hex } from "viem";
import { UNIVERSAL_SIG_VALIDATOR_ADDRESS, universalSigValidatorABI } from "./constants";
import { FacilitatorEvmSigner } from "./signer";

/**
 * Verifies an ERC-6492 counterfactual signature by calling the UniversalSigValidator
 * contract via eth_call (no state changes committed).
 *
 * The validator atomically simulates the factory deployment then verifies the inner
 * signature using EIP-1271 isValidSignature on the resulting contract.
 *
 * @param signer - Facilitator signer for contract reads.
 * @param signerAddress - Address that should have signed.
 * @param hash - EIP-712 typed data hash that was signed.
 * @param signature - Full ERC-6492 wrapped signature bytes.
 * @returns True if the signature is valid, false if invalid or validator unavailable.
 */
export async function verifyERC6492Signature(
  signer: FacilitatorEvmSigner,
  signerAddress: `0x${string}`,
  hash: Hex,
  signature: Hex,
): Promise<boolean> {
  try {
    const result = await signer.readContract({
      address: UNIVERSAL_SIG_VALIDATOR_ADDRESS,
      abi: universalSigValidatorABI,
      functionName: "isValidSig",
      args: [signerAddress, hash, signature],
    });
    if (typeof result !== "boolean") {
      return false;
    }
    return result;
  } catch {
    return false;
  }
}
